extensions/ccui/base-classes/UIScale9Sprite.js

```diff
@@ -576,347 +604,305 @@ ccui.Scale9Sprite = cc.Node.extend(/** @lends ccui.Scale9Sprite# */{
+        this._spriteRect = rect;
+        this._offset = offset;
+        this._spriteFrameRotated = spriteFrameRotated;
+        this._originalSize = size;
+        this._preferredSize = size;
+        this._capInsetsInternal = capInsets;
+        if(this._scale9Enabled)
+            this.createSlicedSprites();
+        else
+            this._scale9Image.initWithTexture(this._scale9Image.getTexture(), this._spriteRect, this._spriteFrameRotated);
 
-        if (this._spritesGenerated) {
-            // Restore color and opacity
+        this.setState(this._brightState);
+        this.setContentSize(size);
+        if(this._spritesGenerated === true) {
             this.setOpacity(opacity);
             this.setColor(color);
         }
         this._spritesGenerated = true;
         return true;
     },
+    /**
+     * Update the scale9Sprite with a SpriteBatchNode.
+     * @param {cc.SpriteBatchNode} batchNode
+     * @param {cc.Rect} originalRect
+     * @param {boolean} rotated
+     * @param {cc.Rect} capInsets
+     * @returns {boolean}
+     */
+    updateWithBatchNode: function (batchNode, originalRect, rotated, capInsets) {
+        var sprite = new cc.Sprite(batchNode.getTexture());
+        var pos = cc.p(0,0);
+        var originalSize = cc.size(originalRect.width,originalRect.height);
+        return this.updateWithSprite(sprite, originalRect, rotated, pos, originalSize, capInsets);
+    },
 
     /**
      * set the sprite frame of ccui.Scale9Sprite
      * @param {cc.SpriteFrame} spriteFrame
+     * @param {cc.rect} capInsets
      */
-    setSpriteFrame: function (spriteFrame) {
-        var batchNode = new cc.SpriteBatchNode(spriteFrame.getTexture(), 9);
-        // the texture is rotated on Canvas render mode, so isRotated always is false.
-        var locLoaded = spriteFrame.textureLoaded();
-        this._textureLoaded = locLoaded;
-        if(!locLoaded){
-            spriteFrame.addEventListener("load", function(sender){
-                // the texture is rotated on Canvas render mode, so isRotated always is false.
-                var preferredSize = this._preferredSize, restorePreferredSize = preferredSize.width !== 0 && preferredSize.height !== 0;
-                if (restorePreferredSize) preferredSize = cc.size(preferredSize.width, preferredSize.height);
-                this.updateWithBatchNode(this._scale9Image, sender.getRect(), cc._renderType === cc._RENDER_TYPE_WEBGL && sender.isRotated(), this._capInsets);
-                if (restorePreferredSize)this.setPreferredSize(preferredSize);
-                this._positionsAreDirty = true;
-                this.dispatchEvent("load");
-            },this);
-        }
-        this.updateWithBatchNode(batchNode, spriteFrame.getRect(), cc._renderType === cc._RENDER_TYPE_WEBGL && spriteFrame.isRotated(), cc.rect(0, 0, 0, 0));
+    setSpriteFrame: function (spriteFrame, capInsets) {
+             var sprite = new cc.Sprite(spriteFrame.getTexture());
```

Comment:
As you removed check for `load` event of texture, if `setSpriteFrame` with a texture not loaded, its size will be all 0, can you test wether it will cause issues ?